FreeRTOS: A few small fixes to make the FreeRTOS port simpler

alistair23 opened the pull request:
This contains two commits that help when porting to FreeRTOS.

alistair23 commented:
Ping!

alistair23 commented:
Ping!

chweimer commented on projects/freertos/cmd_channel_freertos.c:
```diff
@@ -26,7 +25,7 @@ int cmd_channel_freertos_receive_packet (QueueHandle_t rx_queue, struct cmd_pack
 	}
 
-	status = xQueueReceive (rx_queue, packet, timeout);
+	status = xQueueReceive (I2CRequestQueue, packet, timeout);
```
Global variables are typically avoided to allow for flexibility in the implementation.  Having a global variable here prevents this code from being reused in scenarios where you might have multiple command channels.

alistair23 replied:
The current implementation of changing the first argument to
    cmd_channel_freertos_receive_packet()/cmd_channel_freertos_send_packet()
    from a struct *cmd_channel (passed from
    cmd_channel_receive_and_process()) to a QueueHandle_t results in seg faults.
    
    Instead let's create a global queue that can be accessed from anywhere
    to allow data to be stored.
    
    The queue being global allows it to be accessed from interrupt service
    routines (ISRs) which is generally where the data will be processed, at
    least for receives.

alistair23 replied:
What is currently there doesn't work. This is just getting it to work. I agree passing in a queue is better, but that would require changes to the `mctp_cmd_task_loop()` function and a way to handle that in an interrupt handler. I think that should be left for a future patch.

chweimer replied:
It is certainly incorrect to assert what is here doesn't work, because this is actively being using in the current Microsoft implementation.  What I think is missing is a clear example of how these pieces are interrelated, since we don't have a reference port available at this time.

The first thing to point out is that cmd_channel_freertos does not represent an implementation of the cmd_channel API. 
 Perhaps the module could have been named better, but there is no structure of this type defined so it cannot implement the API.  These functions are meant to be helper functions that an implementing cmd_channel could use in a FreeRTOS environment.  However, these functions cannot be directly used as an implementation of the cmd_channel APIs because they do not match the API signature.  There will need to be a wrapper in the cmd_channel implementation that provides the current queue to these calls.

A cmd_channel implementation is expected to be some kind of bus driver (e.g. I2C) that will send and receive packets.  That instance would own Rx and/or Tx FreeRTOS queues that it would pass into these functions as part of receive_packet/send_packet API processing, as appropriate.  Any ISRs would be part of this driver implementation and would have access to the instance structure with the necessary queue(s), so no global piece is required.

alistair23 replied:
Ah, I see.

That is very confusing because there is a `cmd_channel_init()` in `cmd_channel.c` but then `cmd_channel_freertos.c` doesn't match with it.

Would you be open to me adding an example `cmd_channel_receive_packet()` function that matches the `cmd_channel` API and calls `cmd_channel_freertos_receive_packet()`?

chweimer replied:
Sure.  That sounds like a good idea.

alistair23 replied:
Done! Let me know what you think.

chweimer commented on projects/freertos/mctp_cmd_task.c:
```diff
@@ -19,6 +19,7 @@ static void mctp_cmd_task_loop (void *data)
 
 	while (1) {
 		cmd_channel_receive_and_process (task->channel, task->mctp, -1);
+		vTaskDelay(1000);
```
Why is a delay necessary here?

alistair23 replied:
The idea is to change the task to the blocked state to allow other tasks to run.

chweimer replied:
cmd_channel_receive_and_process will call cmd_channel.receive_packet, which will block until a packet is available from the channel.  I see now that the comment on cmd_channel does not make this clear, so I'll plan to update it.

Given that, there is no need to add artificial delay here, and we wouldn't want to prevent another command from being executed until a full second after a previous command, either.

alistair23 committed:
freertos: cmd_channel: Add a default send/receive

Add a default send and receive function that can be used.
Currently cmd_channel_freertos.c includes a receive and send function,
but these can't directly be called from struct cmd_channel. Let's add
default implementations that can be used.

The queue being global allows it to be accessed from interrupt service
routines (ISRs) which is generally where the data will be processed, at
least for receives.

Signed-off-by: Alistair Francis <alistair.francis@wdc.com>

chweimer commented on projects/freertos/cmd_channel_freertos.c:
```diff
@@ -61,3 +61,50 @@ int cmd_channel_freertos_send_packet (QueueHandle_t tx_queue, struct cmd_packet
+void cmd_channel_packet_default_init (struct cmd_channel *channel)
+{
+	channel->receive_packet = cmd_channel_receive_packet;
+	channel->send_packet = cmd_channel_send_packet;
```
If we're looking to turn this into a proper implementation of cmd_channel and do it in way that is consistent with the overall architecture and existing coding style, we would do the following:

1. Define a new 'struct cmd_channel_freertos' that derives from cmd_channel.  This new structure would contain the Rx and Tx queues necessary for receive_packet and send_packet, removing the dependencies on any global context.
2. Create cmd_channel_freertos_init and cmd_channel_freertos_release functions to initialize and free the instance.  The init function returns int (returning error for null parameters), while release returns void (probably with an empty implementation since nothing would be internally created).
3. The init function would take the cmd_channel_freertos instance to initialize and the Rx and Tx queue instances to use with the channel.  It would save these internally and set the function pointers.
4. The functions you have here as cmd_channel_receive_packet/send_packet would be named cmd_channel_freertos_receive_packet/send_packet, which would probably require a name change of the existing functions.  No comment block is necessary on these functions since they are implementations of a base API.
